### PR TITLE
cmd/pin: Update pinned formula messaging

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Resolves https://github.com/Homebrew/brew/issues/16212

The hope is that this will be clearer and less annoying for users.

A user came to us a couple weeks ago stating that it was confusing that the `brew upgrade` command printed an error when a pinned formula had a new version available and didn't get upgraded. Not upgrading a package is the expected result of pinning it.

This PR changes that message to a warning from an error. While looking into this we found that there is another message that gets printed when a package dependent doesn't get upgraded because it is pinned and that got turned into a warning from a normal message. Honestly that should be more worrying for the user anyway; it could lead to a program not working correctly in the worst case scenario.

I also added a message to the `brew pin` command warning about potential unintended behavior if a dependency gets pinned and another package requires a newer version of it.

Lastly, I added a commented out deprecation notice for the `brew upgrade --ignore-pinned` command since it's now essentially the default.

### CODE REVIEW

File: Library/Homebrew/cmd/upgrade.rb [hunk 1]
```diff
@@ -126,6 +126,9 @@ def upgrade_args
   def upgrade
     args = upgrade_args.parse
 
+    # Deprecated since this is now the default behavior.
+    # odeprecated "`brew upgrade --ignore-pinned`" if args.ignore_pinned?
```
Comment (Rylan12) [inline Library/Homebrew/cmd/upgrade.rb:130]: We should set the `--ignore-pinned` switch in the arg parses to `hidden` so it doesn't show up in man pages etc.

File: Library/Homebrew/cmd/upgrade.rb [hunk 2]
```diff
@@ -191,8 +194,8 @@ def upgrade_outdated_formulae(formulae, args:)
       end
     end
 
-    if !pinned.empty? && !args.ignore_pinned?
-      ofail "Not upgrading #{pinned.count} pinned #{Utils.pluralize("package", pinned.count)}:"
+    unless pinned.empty?
```
Comment (Rylan12) [inline Library/Homebrew/cmd/upgrade.rb:197]: Small nit but maybe:

```suggestion
    if pinned.any?
```

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good so far, thanks! My questions are:
1) does this indicates if a specific formula has not been upgraded because of the `brew pin` of a dependency? 
2) does `brew upgrade foo` fail/error if `foo` is pinned?
3) same as 2) but if one of `foo`'s dependencies are pinned.

**Comment by apainintheneck** [comment]
> 1. does this indicates if a specific formula has not been upgraded because of the `brew pin` of a dependency?

> 3. same as 2) but if one of `foo`'s dependencies are pinned.

The answer to 1 and 3 is essentially the same.

If one of foo's dependencies is pinned and the current version is too old, the install will raise an error in the `FormulaInstaller#check_install_sanity` method (great method name by the way).

https://github.com/Homebrew/brew/blob/ab6a7ed5a122b6d4f9d441ef76c86032398ced5f/Library/Homebrew/formula_installer.rb#L348-L352

This error gets caught, printed and then we skip trying to install that formula. That seems to be the case when installing or upgrading formulae so it's even more complex than I originally thought.

https://github.com/Homebrew/brew/blob/ab6a7ed5a122b6d4f9d441ef76c86032398ced5f/Library/Homebrew/install.rb#L280-L282

https://github.com/Homebrew/brew/blob/ab6a7ed5a122b6d4f9d441ef76c86032398ced5f/Library/Homebrew/upgrade.rb#L110-L112

> 2. does `brew upgrade foo` fail/error if `foo` is pinned?

No, it would print a warning. This used to be an error message (ofail) and is the message that was brought up in the original issue.

**Comment by MikeMcQuaid** [comment]
Great, 1 and 3 sound right 👍🏻 

> No, it would print a warning. This used to be an error message (ofail) and is the message that was brought up in the original issue. That means it won't return a non-zero exit code either anymore.

I feel like `brew upgrade` shouldn't print an error or have non-zero exit code but `brew upgrade foo` should because you've specifically requested an upgrade of `foo`, one is available but you've pinned it. Thoughts?

**Comment by apainintheneck** [comment]
> I feel like `brew upgrade` shouldn't print an error or have non-zero exit code but `brew upgrade foo` should because you've specifically requested an upgrade of `foo`, one is available but you've pinned it. Thoughts?

That sounds reasonable. In that case, should we still deprecate the `--ingore-pinned` option? This change probably removes the need for that option for most uses cases but not all.

**Comment by MikeMcQuaid** [comment]
> In that case, should we still deprecate the `--ignore-pinned` option?

I don't feel strongly either way. `brew upgrade foo --ignore-pinned` when `foo` is pinned feels like a pretty weird invocation. If there's other cases outside of that: I'm happy to leave it. Could always leave it for now and see if we end up iterating post-merge based on more feedback?

File: Library/Homebrew/cmd/upgrade.rb [hunk 3]
```diff
@@ -191,8 +195,11 @@ def upgrade_outdated_formulae(formulae, args:)
       end
     end
 
-    if !pinned.empty? && !args.ignore_pinned?
-      ofail "Not upgrading #{pinned.count} pinned #{Utils.pluralize("package", pinned.count)}:"
+    if pinned.any?
+      Kernel.public_send(
+        formulae.any? ? :ofail : :opoo, # only fail when pinned formulae are named explicitly
+        "Not upgrading #{pinned.count} pinned #{Utils.pluralize("package", pinned.count)}:",
+      )
```
Comment (apainintheneck) [inline Library/Homebrew/cmd/upgrade.rb:202]: I handled the case where we want the command to error if a pinned package is explicitly passed to `brew upgrade` but not when it's called without package arguments (https://github.com/Homebrew/brew/pull/16301/commits/9a00f2795d9eca48385a5222d953787b0269f050).

```console
$ brew upgrade
Warning: Not upgrading 1 pinned package:
fish 3.6.4
==> No packages to upgrade
$ brew upgrade fish
Error: Not upgrading 1 pinned package:
fish 3.6.4
==> No packages to upgrade
```

**Comment by apainintheneck** [comment]
> I don't feel strongly either way. brew upgrade foo --ignore-pinned when foo is pinned feels like a pretty weird invocation.

Me thoughts exactly. I'll leave the deprecation in.

File: Library/Homebrew/cmd/upgrade.rb [hunk 4]
```diff
@@ -126,6 +127,9 @@ def upgrade_args
   def upgrade
     args = upgrade_args.parse
 
+    # Deprecated since this is now the default behavior.
+    # odeprecated "`brew upgrade --ignore-pinned`" if args.ignore_pinned?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/upgrade.rb:131]:
```suggestion
    odeprecated "`brew upgrade --ignore-pinned`" if args.ignore_pinned?
```

Deprecating this immediately as new tag will be a minor release 🎉

**Review by MikeMcQuaid — APPROVED** [review]
Thanks for all the work here @apainintheneck and sorry for the amount of back and forth needed!

**Comment by Moulick** [comment]
> > In that case, should we still deprecate the `--ignore-pinned` option?
> 
> I don't feel strongly either way. `brew upgrade foo --ignore-pinned` when `foo` is pinned feels like a pretty weird invocation. If there's other cases outside of that: I'm happy to leave it. Could always leave it for now and see if we end up iterating post-merge based on more feedback?

Does feel weird, the one case that comes to mind is that if `foo` is pinned and you want to upgrade `foo` only this time but leave it pinned for the next time, without having to unpin, upgrade, re-pin

**Comment by apainintheneck** [comment]
> Does feel weird, the one case that comes to mind is that if foo is pinned and you want to upgrade foo only this time but leave it pinned for the next time, without having to unpin, upgrade, re-pin

@Moulick I don't believe that was possible before this change either. This strictly changes messaging not how `brew install` behaves when a formula is pinned.

**Comment by Moulick** [comment]
> > Does feel weird, the one case that comes to mind is that if foo is pinned and you want to upgrade foo only this time but leave it pinned for the next time, without having to unpin, upgrade, re-pin
> 
> @Moulick I don't believe that was possible before this change either. This strictly changes messaging not how `brew install` behaves when a formula is pinned.

Sure, I get that this is just a message change. From a power users' perspective, looking at the command `brew upgrade foo --ignore-pinned`, logically I'd think it would upgrade foo ignoring the fact that it is pinned or not, even though it does not do that. Anyways, I don't think we really need it anyways, it's more of a convenience that would help if someone wanted to upgrade without having to unpin, but for now a user can unpin, upgrade, re-pin if they so wanted ¯\_(ツ)_/¯

**Pull request merged**